### PR TITLE
Add token_count client method

### PR DESCRIPTION
**Pull request by jamiemccarthy**
## token_count

As discussed in #291 , it can be useful in many cases to support calculating the token-count of messages before sending them to OpenAI.

This PR adds an `OpenAI::Client.token_count` instance method which does this. To minimize astonishment, it accepts the same `parameters` input as the other methods. Its output is always a nonnegative integer.

If the caller has already `require`'d the `tiktoken_ruby` module, this new method automatically uses it to calculate exactly the number of tokens that OpenAI will parse. (Because `parameters` includes the model name, we know we're using the same algorithm OpenAI will use.)

Or, if `Tiktoken` is not available, it uses the two "helpful rules of thumb" in OpenAI's [docs](https://help.openai.com/en/articles/4936856-what-are-tokens-and-how-to-count-them) and averages them as an estimate.

Because testing the `Tiktoken` code paths works best if the module is actually included and not mocked, this PR adds that gem to the `Gemfile`. Note that this does not change the dependency list. Also, because the test is written against two specific, existing models, that module's counts should not change in the future and thus the test results should be pretty future-proof.

I'm happy to make any changes desired. Just let me know.

## ruby 2.6

`Tiktoken` requires ruby >= 2.7.  This gem requires ruby >= 2.6. I wanted to keep this gem's test suite passing on all versions of ruby.

What I did was the slightly unorthodox method of editing `Gemfile` to only make mention of `tiktoken` when running on ruby >= 2.7, and add conditionals in the test file to skip those three tests on ruby 2.6.

It works — I checked it locally on ruby 2.6, 2.7, and 3.1.

But... it's a little ugly. Personally, if you asked me, since 2.6 isn't getting security updates, I wouldn't want to encourage its use. I would revert the ugly changes (f8d548103a6f11b8231192894f2951fa3e6de5ef) and set the [required version](https://github.com/alexrudall/ruby-openai/blob/ef982fe99ed3585c2238a33ab891294113fe894a/ruby-openai.gemspec#L12) to 2.7. But I didn't want to make that change without discussing/consulting about it first. What's in this PR works fine and changes things as minimally as possible.

## All Submissions:

* [x] Have you followed the guidelines in our [Contributing document](../blob/main/CONTRIBUTING.md)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../pulls) for the same update/change?
* [x] Have you added an explanation of what your changes do and why you'd like us to include them?

### CODE REVIEW

File: lib/openai/client.rb [hunk 1]
```diff
@@ -54,5 +54,32 @@ def transcribe(parameters: {})
     def translate(parameters: {})
       OpenAI::Client.multipart_post(path: "/audio/translations", parameters: parameters)
     end
+
+    def token_count(parameters: {})
```
Comment (codenamev) [inline lib/openai/client.rb:58]: This feels strange to me living inside the `Client` class.  I'd expect public methods in this context to interface with OpenAI in some way.  Consider moving this to a new utility class or included on the top-level `OpenAI` namespace.

File: lib/openai/client.rb [hunk 2]
```diff
@@ -54,5 +54,32 @@ def transcribe(parameters: {})
...
+      contents = parameters[:messages]&.map { |message| message[:content] }&.join(" ") || nil
```
Comment (codenamev) [inline lib/openai/client.rb:59]: I'd like to vote against forcing this method to expect multiple messages by default. Consider using a single `content = ""` parameter and reducing the complexity here.

Comment (alexrudall) [reply lib/openai/client.rb:59]: Agree with this

File: spec/openai/client/token_count_spec.rb [hunk 3]
```diff
@@ -0,0 +1,60 @@
+RSpec.describe OpenAI::Client do
+  describe "#token_count" do
+    let(:messages) { [
+      { content: "Red is my favorite color." },
+      { content: "You miss 100% of the shots you don't take." },
+      { content: "Ingredients for banana bread can include 1 egg or 3 eggs." },
+      { content: "Egg is not a necessary ingredient." }
```
Comment (alexrudall) [inline spec/openai/client/token_count_spec.rb:7]: 🥚

**Comment by jamiemccarthy** [comment]
> * Rename the method to `rough_token_count`

I'm happy to do this, but I'd find it less useful. Does it change your mind to know that OpenAI's recommended estimation algorithms are usually high by 10-20%?

To me, that's a lot of wiggle room for error, if people are trying to have their code make decisions based on that estimate.

For example, in the largest example given in [the docs](https://help.openai.com/en/articles/4936856-what-are-tokens-and-how-to-count-them), the Declaration of Independence, it's off from Tiktoken's calculations by 13-17%:

```
jamie@Jamies-MacBook-Air ruby-openai % bundle exec irb                                         
irb(main):001:0> require 'openai'
=> true
irb(main):002:0> OpenAI::Client.new.token_count(parameters: { model: "tktk", messages: [ { content: File.read("dofi")}]})
=> 1887
irb(main):003:0> require 'tiktoken_ruby'
=> true
irb(main):004:0> Tiktoken.encoding_for_model("gpt-4").encode(File.read("dofi")).size
=> 1616
irb(main):005:0> Tiktoken.encoding_for_model("davinci").encode(File.read("dofi")).size
=> 1674
```

**Comment by jamiemccarthy** [comment]
Thanks for your feedback. I've submitted #312 which I believe addresses it.

**Pull request closed without merging**